GR_LYCHEE,RZ_A1H,VK_RZ_A1H: Update to fix ARMC6 build failures

d-kato opened the pull request:
### Description
Fix for issue #9563

We reviewed the compile switch.
Because the rules of inline assembler are different, we changed part of code to assembler code.

When building ARMV6, the "-mthumb" option is specified. And "rtx_core_ca.h" had a problem not working in Thumb.
https://github.com/ARM-software/CMSIS_5/issues/526


If either of the following is merged with this PR, this problem will be solved.
https://github.com/ARM-software/CMSIS_5/pull/527
https://github.com/ARMmbed/mbed-os/pull/9612

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

d-kato committed:
GR_LYCHEE,RZ_A1H,VK_RZ_A1H: Update to fix ARMC6 build failures

d-kato commented on rtos/TARGET_CORTEX/rtx5/RTX/Source/rtx_core_ca.h:
```diff
@@ -112,6 +112,18 @@ static __asm    uint32_t __get_PSP (void) {
+    "sub  r0,r0,#32\n\t"
+    "bx   lr\n\t"
+  );
+}
```
In the conventional writing style, "arm" and "thumb" are mixed when "-mthumb" option is specified, and it does not work properly.
I wrote the whole function including the return value with the inline assembler.

kjbracey replied:
This doesn't seem right for a few reasons.

* It's a change to upstream CMSIS-RTOS code
* It doesn't put the assembler state back to Thumb afterwards, if it was
* It doesn't mark the function itself as being ARM, rather than Thumb, so you'll be entered in Thumb state, if `-mthumb` is enabled. The assembler is being switched, not the CPU.

Now, looking at the original CMSIS-RTOS code, I think this header file doesn't actually cope with the compiler being set to Thumb at all for GCC or ARMC5, and that's a pre-existing issue - it's assuming the compiler is set to compile ARM.

I think we should probably side-step the issue for now by not attempting to use Thumb mode when compiling for A-profile. ARMC6 is out of line here. GCC and ARMC5 are both set to use ARM, and that's why this is failing for ARMC6.

I think the `-mthumb` should be removed from the ARMC6 profiles. I don't think you need to specify at all to match the other toolchains - the default is `-marm` if specifying a CPU that supports it, else `-mthumb`. If it did need specification, then it should be in `arm.py` like it is in `gcc.py`, where it's set based on CPU.

kjbracey replied:
Raising a separate Mbed OS PR to change the tool profiles (#9612). Someone could submit a CMSIS-RTOS issue to make this header more flexible.

d-kato replied:
Thank you for review. I reverted it.

kjbracey replied:
It would be useful if you could confirm that each of https://github.com/ARMmbed/mbed-os/pull/9612 and https://github.com/ARM-software/CMSIS_5/pull/527 solve the issue on their own.

ciarmcom commented:
@d-kato, thank you for your changes.
@ARMmbed/mbed-os-maintainers please review.

d-kato committed:
Back to the original : "rtx_core_ca.h"

kjbracey committed:
RTX5: Make rtx_core_ca.h cope with compiler in Thumb mode

Assembler in rtx_core_ca.h is on the whole unified syntax such that it
should work fine if assembled as either ARM or Thumb-2.

The exception was __get_PSP, which uses STM^, which is only available
in ARM state.

Flagging for this function was incorrect, except for IAR - it switched
assembler state without telling the compiler, meaning that the resulting
assembler output could be incorrect, and that the function itself would
not be correctly marked as an ARM function - the CPU would enter in
Thumb state.

Alternative fix would be to switch to System mode, which would work
as either ARM or Thumb-2 assembler, like the rest of the file, but
this is the minimal change.

Fixes #526.

NirSonnenschein commented:
@d-kato  please add a description of the issue and fix.

cmonr commented:
> https://github.com/ARMmbed/mbed-os/pull/9609#issuecomment-461020649

@NirSonnenschein What was this in reference to?

0xc0170 commented:
Can you rebase ? there's 9430 included, should not be here

0xc0170 commented:
@d-kato As you suggested changes for #9612 , how to address that one and this PR?

d-kato commented:
OK, I will rebase.
Since ARMC6 builds in Thumb, https://github.com/ARMmbed/mbed-os/pull/9612 is not necessary. Instead, we need to change https://github.com/ARM-software/CMSIS_5/pull/527, should I add it to this PR?

kjbracey commented:
I'm going to vote for yes, go ahead and add the CMSIS change. Make sure to apply it carefully to avoid future conflicts.

(When I'm doing it I have a clone of the CMSIS_5 repository, I use `git format-patch` to extract the commit in patch form, edit the patch file to modify the paths, then use `git am` to apply to the Mbed OS tree. Edit the commit message to link to the CMSIS_5 commit ID and/or PR. If anyone has a smoother operation, do shout).

cmonr commented:
> If anyone has a smoother operation, do shout).

@deepikabhavnani any thoughts to @kjbracey-arm's comment/request?

deepikabhavnani commented:
I don't have better suggestion for particular patch, but if overall CMSIS update is needed we can use importer tools https://github.com/ARMmbed/mbed-os/tree/master/tools/importer

In case of local updates, make sure to cross check the commits listed here: https://github.com/ARMmbed/mbed-os/blob/master/tools/importer/cmsis_importer.json#L146 are not broken

kjbracey commented:
Don't think we need a global update at this point; this patch should apply fine - it's to the core_a.h files we don't usually use.
